data_structures/heap/heap.py

```diff
@@ -1,99 +1,131 @@
+            # check which child is larger than its parent
+            if left_child is not None and self.h[left_child] > self.h[violation]:
+                violation = left_child
+            if right_child is not None and self.h[right_child] > self.h[violation]:
+                violation = right_child
+            # if violation indeed exists
+            if violation != index:
+                # swap to fix the violation
+                self.h[violation], self.h[index] = self.h[index], self.h[violation]
+                # fix the subsequent violation recursively if any
+                self.max_heapify(violation)
+
+    def build_max_heap(self, collection: Iterable[Union[int, float]]):
+        """ build max heap from an unsorted array"""
         self.h = list(collection)
-        if self.curr_size <= 1:
-            return
-        for i in range(self.curr_size // 2 - 1, -1, -1):
-            self.max_heapify(i)
-
-    def get_max(self):
-        if self.curr_size >= 2:
+        self.heap_size = len(self.h)
+        if self.heap_size > 1:
+            # max_heapify from right to left but exclude leaves (last level)
+            for i in range(self.heap_size // 2 - 1, -1, -1):
+                self.max_heapify(i)
+
+    def max(self) -> Union[int, float]:
+        """ return the max in the heap """
+        if self.heap_size >= 1:
+            return self.h[0]
+        else:
+            raise Exception("Empty heap")
+
+    def extract_max(self) -> Union[int, float]:
+        """ get and remove max from heap """
+        if self.heap_size >= 2:
             me = self.h[0]
             self.h[0] = self.h.pop(-1)
-            self.curr_size -= 1
+            self.heap_size -= 1
             self.max_heapify(0)
             return me
-        elif self.curr_size == 1:
-            self.curr_size -= 1
+        elif self.heap_size == 1:
+            self.heap_size -= 1
             return self.h.pop(-1)
-        return None
+        else:
+            raise Exception("Empty heap")
+
+    def insert(self, value: Union[int, float]):
+        """ insert a new value into the max heap """
+        self.h.append(value)
+        idx = (self.heap_size - 1) // 2
+        self.heap_size += 1
+        while idx >= 0:
+            self.max_heapify(idx)
+            idx = (idx - 1) // 2
 
     def heap_sort(self):
-        size = self.curr_size
+        size = self.heap_size
         for j in range(size - 1, 0, -1):
             self.h[0], self.h[j] = self.h[j], self.h[0]
-            self.curr_size -= 1
+            self.heap_size -= 1
             self.max_heapify(0)
-        self.curr_size = size
-
-    def insert(self, data):
-        self.h.append(data)
-        curr = (self.curr_size - 1) // 2
-        self.curr_size += 1
-        while curr >= 0:
-            self.max_heapify(curr)
-            curr = (curr - 1) // 2
+        self.heap_size = size
 
-    def display(self):
-        print(self.h)
 
-
-def main():
+def demo():
```

Comment:
Please move the function content in the `__name__ == "__main__":` block, not as a function but as statemens only.